kernel/aster-nix/src/ipc/mod.rs

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use crate::{
+    prelude::*,
+    process::{Gid, Uid},
+};
+
+pub mod semaphore;
+
+#[allow(non_camel_case_types)]
+pub type key_t = i32;
+
+bitflags! {
+    pub struct IpcFlags: u32{
+        /// Create key if key does not exist
+        const IPC_CREAT  = 1 << 9;
+        /// Fail if key exists
+        const IPC_EXCL  = 1 << 10;
+        /// Return error on wait
+        const IPC_NOWAIT  = 1 << 11;
+        /// Undo the operation on exit
+        const SEM_UNDO = 1 << 12;
+    }
+}
+
+#[repr(i32)]
+#[derive(Debug, Clone, Copy, TryFromInt)]
+#[allow(non_camel_case_types)]
+pub enum IpcControlCmd {
+    IPC_RMID = 0,
+    IPC_SET = 1,
+    IPC_STAT = 2,
+
+    SEM_GETPID = 11,
+    SEM_GETVAL = 12,
+    SEM_GETALL = 13,
+    SEM_GETNCNT = 14,
+    SEM_GETZCNT = 15,
+    SEM_SETVAL = 16,
+    SEM_SETALL = 17,
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct IpcPermission {
+    key: key_t,
+    /// Owner's UID
+    uid: Uid,
+    /// Owner's GID
+    gid: Gid,
+    /// Creator's UID
+    cuid: Uid,
+    /// Creator's GID
+    cguid: Gid,
+    /// Permission mode
+    mode: u16,
```

Comment:
Using `PermissionMode` is better than u16, if possible

Reply:
I am not sure if other IPCs like POSIX semaphores will follow the definition of the `PermissionMode` in System V semaphore. So I keep the u16 here.

It may be better to use `PermissionMode` when implementing permission checks in the future PR.